### PR TITLE
Allowing non-GET-verb URL rules to have a URL created

### PR DESCRIPTION
**Pull request by bizley**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | only if someone depends on non-GET rule URL to not be created
| Tests pass?   | ✔️
| Fixed issues  | #18487

### CODE REVIEW

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ Yii Framework 2 Change Log
 - Bug #18479: Fix invalid argument type for `preg_split` in `\yii\console\Controller` (gazooz)
 - Bug #18477: Fix detecting availability of Xdebug's stack trace in `yii\base\ErrorException` (bizley)
 - Bug #18480: Transactions are not committed using the dblib driver (bbrunekreeft)
+- Enh #18487: Allowing non-GET-verb URL rules to have a URL created (bizley)
```
Comment (samdark) [inline framework/CHANGELOG.md:17]:
```suggestion
- Enh #18487: Allow creating URLs for non-GET-verb rules (bizley)
```

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Good one! Thanks!

**Comment by rob006** [comment]
This is BC breaking change, it should at least be mentioned in UPGRADE.md. Also old behavior is still described in documentation: https://github.com/yiisoft/yii2/blob/88e79d3f8eb0121542d8174762cc9a1dca35313d/docs/guide/runtime-routing.md#http-methods-

Personally I don't think that changing this old and documented behavior is the right approach. I would rather allow to set explicitly that rule should be used for URL creation regardless of verbs settings. For `yii\web\UrlRule` this should already be possible by using array syntax (`['pattern' =>'...', 'route' => '...', ...]`) for rule definition (which is a little surprising and inconsistent). Maybe `yii\rest\UrlRule` should support array syntax too.

**Comment by bizley** [comment]
Yes, I was thinking about adding it to upgrade. Will do. Also I was looking for this in the docs and could not find, cheers.

As for the second part - are you suggesting there should be some additional option to allow that explicitly? To this time using array syntax with `['pattern' => 'POST x', ...]` would exclude the rule from URL creating so it's not just the array syntax thing. And what do you mean by rest rules using array syntax?

**Comment by samdark** [comment]
Mentioning it is OK but I can't imagine a real-enough situation where it may cause a problem.

**Comment by rob006** [comment]
> To this time using array syntax with `['pattern' => 'POST x', ...]` would exclude the rule from URL creating

I'm pretty sure it won't, since verbs were processed from pattern only if `is_string($rule)` returns `true`, so verb should not be even extracted from `['pattern' => 'POST x', ...]` - `'POST x'` will be treated as literal pattern. Rule defined as `['pattern' => 'login', 'route' => 'login/login', 'verb' => 'POST']` should be used also for URL creation before this PR. If we support the same syntax for `yii\rest\UrlRule::$extraPatterns` and `yii\rest\UrlRule::$patterns`, there will be no need to BC break.

**Comment by bizley** [comment]
Oh, you are right, my bad.

**Comment by bizley** [comment]
I'm not sure though what solution you are proposing. You are right that `['pattern' => 'login', 'route' => 'login/login', 'verb' => 'POST']` is allowed for creation (at least for non-rest rules) but `['POST login' => 'login/login']` not. Isn't the removing of such implementation a good thing?

**Comment by rob006** [comment]
It was documented and worked in that way since 7 years, I don't think that changing this right now is a wise move. I would rather treat `['POST login' => 'login/login']` as shortcut with some magic conventions attached, and `['pattern' => 'login', 'route' => 'login/login', 'verb' => 'POST']` explicit syntax that allows you skip this magic (for cases like #18487). Instead of changing this, I would rather extend this pattern to `yii\rest\UrlRule::$extraPatterns` and `yii\rest\UrlRule::$patterns` - it will improve consistency without BC break.

@samdark Isn't the whole point of involving verbs in URL parsing/creation is to differentiate URLs depending on verb? In that case every single case where defining verb for rule make sense will be affected by this change.

**Comment by bizley** [comment]
Up to 2.0.40 this is allowed for creating URLs:

```php
[
    'class' => 'yii\web\UrlRule`,
    'pattern' => 'POST user',
    'route' => 'user/create',
]
```

but this

```php
['POST user' => 'user/create']
```

and this

```php
[
    'class' => 'yii\rest\UrlRule`,
    'controller' => 'user',
    'pluralize' => false,
    'only' => ['create'], // or 'patterns' => ['POST' => 'create']
]
```

are **not**. At the same time we can see these are basically the same in terms of request parsing. This is inconsistent.

Also for the second part:

For classic REST rule for `user` controller and `POST`, `GET`, `HEAD`, and `OPTIONS` verb `createUrl()` will result with the same `users`. When sending the request all depends on the sent verb so it will be routed to the proper action, just like before. But now 4 verbs can be used to create the URL to be sent nad not just the only one.

I cannot imagine a case where it can be a problem (but hey, I was wrong before so... ;)

**Comment by rob006** [comment]
> Up to 2.0.40 this is allowed for creating URLs:
> 
> ```
> [
>     'class' => 'yii\web\UrlRule`,
>     'pattern' => 'POST user',
>     'route' => 'user/create',
> ]
> ```

Are you sure that ` 'pattern' => 'POST user',`  will work that way? I can't find where verb is extracted from `pattern` key. It is only handled in case of `['POST user' => 'user/create']` syntax. And in that case it looks consistent if you thing more about it - `['POST user' => 'user/create']` is "magic" syntax that extract verb, pattern and sets mode based on key. And none of this happens when you pass UrlRule config as array - you need to set everything explicitly. 

> But now 4 verbs can be used to create the URL to be sent nad not just the only one.

So for failures it works 4 times slower (4 rules needs to be processed while creating URL that does nit match specific rule - previously only one could be used for creating URLs)? :D

**Comment by bizley** [comment]
Argh, right, it should be 

```php
[
    'class' => 'yii\web\UrlRule`,
    'pattern' => 'user',
    'verb' => 'POST',
    'route' => 'user/create',
]
```

but all other is true. And it does not work 4 times slower, see my [benchmark](https://github.com/yiisoft/yii2/issues/18487#issuecomment-766772070).

**Pull request merged**